### PR TITLE
Add rule result sample ingestion metrics.

### LINKED ISSUE #8386: Ruler manager doesn't reflect `appender.Commit()` errors in the metrics
## Proposal

**Use case**:

Investigating a Cortex ruler problem recently. I noticed that the ruler manager doesn't reflect errors from `appender.Commit` in the ruler evaluation metrics:
https://github.com/prometheus/prometheus/blob/b7fe028740b7b36a31c2deda1e2b74aa566fc0ee/rules/manager.go#L618

While the errors are handled for `appender.Add` ( and reflected correctly in the metrics there is no accounting for errors that happen on  `appender.Commit`:
https://github.com/prometheus/prometheus/blob/b7fe028740b7b36a31c2deda1e2b74aa566fc0ee/rules/manager.go#L625

The Cortex  `Appendable` isn't able to verify problems with the samples during `Add()`, so out-of-order or duplicate metrics will only fail at `Commit` time.

**Proposal**:

Errors during a `Commit` should be accounted for with a similar logic as for `Add`, the relevant error counter needs to be increased by the value of successful `Add` calls

### PR DESCRIPTION
**Pull request by simonswine**
This adds two metrics `rule_result_ingestion_samples_total` and `rule_result_ingestion_sample_failures_total` to surface rule result ingestion failures.

Fixes #8386

### CODE REVIEW

File: rules/manager.go [hunk 1]
```diff
@@ -115,6 +117,22 @@ func NewGroupMetrics(reg prometheus.Registerer) *Metrics {
 			},
 			[]string{"rule_group"},
 		),
+		ingestionSamplesTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Name:      "rule_result_ingestion_samples_total",
```
Comment (brian-brazil) [inline rules/manager.go:123]: rule_group_ingested_samples_total would be consistent with the existing metrics. The help should make clear that this is samples we attempted to ingest, as distinct from what was actually ingested.

File: rules/manager.go [hunk 2]
```diff
@@ -115,6 +117,22 @@ func NewGroupMetrics(reg prometheus.Registerer) *Metrics {
...
+				Help:      "The total number of result samples generated for ingestion.",
+			},
+			[]string{"rule_group"},
+		),
+		ingestionFailuresTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Namespace: namespace,
+				Name:      "rule_result_ingestion_sample_failures_total",
```
Comment (brian-brazil) [inline rules/manager.go:131]: There should be a metric per failure type, to match scraping.

File: rules/manager.go [hunk 3]
```diff
@@ -616,13 +638,16 @@ func (g *Group) Eval(ctx context.Context, ts time.Time) {
...
 			defer func() {
 				if err := app.Commit(); err != nil {
+					g.metrics.ingestionFailuresTotal.WithLabelValues(groupKey(g.File(), g.Name())).Add(float64(len(seriesReturned)))
 					level.Warn(g.logger).Log("msg", "Rule sample appending failed", "err", err)
 					return
 				}
 				g.seriesInPreviousEval[i] = seriesReturned
 			}()
 			for _, s := range vector {
+				g.metrics.ingestionSamplesTotal.WithLabelValues(groupKey(g.File(), g.Name())).Inc()
```
Comment (brian-brazil) [inline rules/manager.go:648]: It'd be better to do these as one increment after the for.

**Comment by roidelapluie** [comment]
Hello @simonswine , did you see the comments? It also seems that there are conflicts now. Thanks!!

File: rules/manager.go (hunk 2 above)
Comment (simonswine) [reply rules/manager.go]: I have tried to follow scraping as much as possible, but there needs to be a catch all for errors that don't match the other categories which I called `other_failures`

File: rules/manager.go (hunk 3 above)
Comment (simonswine) [reply rules/manager.go]: Done that :+1:

File: rules/manager.go (hunk 1 above)
Comment (simonswine) [reply rules/manager.go]: > Help:      "Total number of ruler result samples attempted to ingest."

I guess the ingested certainly implies that an ingestion has been successful (at least to me). I have used never the less in the last commit.

Would `s/prometheus_rule_group_ingested/rule_group_ingestion_samples_total` make it clearer?

**Comment by simonswine** [comment]
@roidelapluie finally got round to rebase this and address the comments. I am not 100% happy with the metric names yet, any suggestions would be very welcome :slightly_smiling_face:

**Comment by gouthamve** [comment]
I am not sure I like this approach tbh, to monitor if my rules are suceeding, I would now need to monitor 4 different metrics (eval fails, 3x append fails).

Can we simplify this to a single metric for all the cases (eval + append fails) or atleast a single metric for the append fail?

**Comment by roidelapluie** [comment]
yeah, I did not have the time to review this deeply but in my setups this would be too high cardinality. it's a lot of metrics and I have a lot of groups (and so do our users): https://github.com/prometheus/prometheus/issues/8548

**Comment by simonswine** [comment]
Closed in favor of #8770

**Pull request closed without merging**